Model to trace to sweep automation

Aswinmcw opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/29451

README : https://github.com/tenstorrent/tt-metal/blob/Aswinmcw/model_trace/model_tracer/README.md

### Problem Description

**Problem: Sweep tests don't necessarily contain configurations from the models we purport to support**

Current sweep tests are manually hardcoded configurations that may not reflect actual model usage. This creates several issues:

1. **Configuration Gap**: Test configurations are arbitrary and may miss cases that occur in production models (EfficientNet, ResNet, BERT, etc.)

2. **Manual Maintenance**: Developers must manually identify and add representative configurations

3. **Sharded Memory Configs**: Sharded memory configurations (HEIGHT_SHARDED, WIDTH_SHARDED, BLOCK_SHARDED) with specific shard shapes, core grids, and orientations are difficult to create manually

## What's Changed

**Solution: Automated sweep test creation from real models via traced configurations**

This system automatically extracts and creates operation configurations used in actual models:

1. **Automatic Tracing**: Captures all TTNN operation configurations from model test execution (shapes, dtypes, layouts, exact shard specs) using graph capture

2. **Centralized Storage**: Deduplicates and stores unique configurations in `ttnn_operations_master.json` across all traced models

3. **Drop-in Integration**: 2-line code change to add `model_traced` suite to any sweep test - automatically loads and uses real configurations

4. **Exact Replication**: Preserves complete memory configurations including shard_shape, core grid ranges, and orientation from production models

5. **Continuous Validation**: As new models are traced, their configurations are automatically added to the test suite, ensuring ongoing validation

6. **Dashboard Visibility**: Test results appear in Superset dashboard showing pass/fail rates for model-traced configurations vs synthetic ones

**Impact**: Sweep tests now validate operations that are used in production models

---

## System Workflow

```mermaid
flowchart TB
    Start([Model Test File]) --> Tracer[🔍 generic_ops_tracer.py]
    
    Tracer --> |Captures Operations| Graph[TTNN Graph Trace]
    Graph --> |Filters Valid Ops| Filter[Filter by Allops.txt]
    Filter --> |Extracts Configs| Extract[Extract Shapes, Dtypes,<br/>Memory Configs, Layouts]
    
    Extract --> Master[(📦 Master JSON<br/>ttnn_operations_master.json)]
    
    Master --> |Deduplicated<br/>Unique Configs| Master
    
    Master -.->|Query| Analyze[🔎 analyze_operations.py<br/>View Configs]
    
    Master --> Loader[⚙️ master_config_loader.py<br/>Parse & Convert]
    
    Loader --> |Provides Parameters| Sweep[📊 Sweep Test<br/>model_traced suite]
    
    Sweep --> |Generate Vectors| VectorGen[sweeps_parameter_generator.py]
    VectorGen --> Vectors[(Test Vectors)]
    
    Vectors --> Runner[🚀 sweeps_runner.py]
    Runner --> |Execute Tests| Device[🔲 TTNN Device]
    Device --> |Results| Results[(✅ Test Results)]
    
    Results --> Dashboard[📈 Superset Dashboard<br/>Pass/Fail Analytics]
    
    style Start fill:#e1f5ff
    style Master fill:#fff4e1
    style Sweep fill:#e8f5e9
    style Dashboard fill:#f3e5f5
    style Tracer fill:#ffebee
    style Loader fill:#fff9c4
    style Runner fill:#e0f2f1
```

**Key Flow:**
1. 🔍 **Trace** → Extract ops from model tests
2. 📦 **Store** → Deduplicate & save to master JSON
3. ⚙️ **Load** → Parse configs into sweep parameters
4. 📊 **Test** → Run sweep tests with real parameters
5. 📈 **Analyze** → View results in dashboard

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

Aswinmcw committed:
Model to trace to sweep automation

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Add comprehensive guide for TTNN Traced Configurations System and master config loader

### What's changed
- Introduced `TRACED_CONFIGS_GUIDE.md` detailing the TTNN Traced Configurations System, including usage, architecture, and examples.
- Added `master_config_loader.py` to facilitate loading and parsing of operation configurations from the master JSON file for sweep tests.

### Benefits
- Provides a centralized resource for understanding and utilizing the traced configurations.
- Enhances integration of real-world operation configurations into sweep tests, improving testing accuracy and efficiency.

Aswinmcw committed:
Add operation analyzer script and update configuration guide

### What's changed
- Introduced `analyze_operations.py`, a script for analyzing the `ttnn_operations_master.json` file, allowing users to view configurations, search operations, and display statistics.
- Updated `TRACED_CONFIGS_GUIDE.md` to include instructions for using the new analyzer script and simplified integration steps for sweep tests.
- Refactored `master_config_loader.py` to support fetching operation parameters for sweep tests, enhancing usability and performance.

### Benefits
- Provides a comprehensive tool for analyzing and managing TTNN operation configurations.
- Streamlines the integration of traced configurations into testing workflows, improving testing accuracy and efficiency.

Aswinmcw committed:
Add sweep results analyzer script and enhance configuration guide

### What's changed
- Introduced `analyze_sweep_results.py`, a script for analyzing sweep test result JSON files, providing detailed statistics and performance metrics.
- Updated `TRACED_CONFIGS_GUIDE.md` to include integration patterns for binary operations and usage of the new analyzer script.
- Enhanced `master_config_loader.py` to support detection of binary operations and streamline parameter loading for sweep tests.

### Benefits
- Offers a comprehensive tool for analyzing sweep test results, improving visibility into test performance and outcomes.
- Simplifies the integration of binary operation configurations into testing workflows, enhancing testing accuracy and efficiency.

Aswinmcw committed:
Update TRACED_CONFIGS_GUIDE.md to reflect new system name and description

### What's changed
- Renamed the guide from "TTNN Traced Configurations System" to "Model Trace to Test Automation".
- Updated the description to align with the new system name while maintaining the original content's intent.

### Benefits
- Clarifies the purpose of the guide, ensuring it accurately represents the system's functionality and scope.

Aswinmcw committed:
update

Aswinmcw committed:
update

Aswinmcw committed:
update

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
update

Aswinmcw committed:
update

Aswinmcw committed:
Add model tracing and analysis tools for enhanced testing automation

### What's changed
- Introduced `analyze_operations.py` for analyzing TTNN operation configurations, allowing users to view, search, and display statistics.
- Added `analyze_sweep_results.py` for comprehensive analysis of sweep test results, including pass/fail statistics and performance metrics.
- Implemented `generic_ops_tracer.py` to extract operations from model tests, facilitating automated configuration generation.
- Created `TRACED_CONFIGS_GUIDE.md` to provide detailed documentation on the new system, including usage examples and integration patterns.
- Updated `master_config_loader.py` to support the new file structure for loading traced configurations.

### Benefits
- Streamlines the process of integrating real-world model configurations into sweep tests, improving testing accuracy and efficiency.
- Provides robust tools for analyzing both operation configurations and sweep test results, enhancing visibility into testing performance.

Aswinmcw committed:
update

Aswinmcw committed:
update

Aswinmcw committed:
update

Aswinmcw committed:
update

cmaryanTT commented on tests/sweep_framework/sweeps/eltwise/binary/add/add_all_pytorch2.py:
Let's move the new content out of the pytorch2 file, and into a new file "add_model_trace.py" or similar. Keep the pytorch2 traces separate from this effort.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Enhance generic_ops_tracer with argparse support and trace file management

- Added argparse for improved command-line interface, allowing users to specify test paths and output directories.
- Introduced `--store` flag to retain individual trace files after execution, with cleanup logic for unused files.
- Updated documentation to reflect new options and usage examples for the tracer.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Remove hardcoded environment setup from generic_ops_tracer.py

Aswinmcw committed:
Add SPDX license headers to analyze_operations.py, analyze_sweep_results.py, generic_ops_tracer.py, and master_config_loader.py

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Update TRACED_CONFIGS_GUIDE.md for clarity and structure

- Renamed the guide to "Model Trace to Test Automation Guide" for better clarity.
- Streamlined the table of contents and added an overview section highlighting the benefits of the system.
- Improved integration instructions for unary and binary operations, emphasizing simplicity and clarity.
- Enhanced examples and usage sections for better understanding of the tracing and testing process.
- Removed redundant content and ensured consistent formatting throughout the document.

Aswinmcw committed:
Remove outdated troubleshooting section from TRACED_CONFIGS_GUIDE.md to streamline content and improve clarity.

Aswinmcw commented:
@stevendae can you please take a look at this PR?

patrickroberts commented on tests/sweep_framework/sweeps/eltwise/unary/sigmoid_accurate/sigmoid_accurate.py:
```diff
@@ -41,22 +56,44 @@
+    if input_a_memory_config is None:
+        input_a_memory_config = ttnn.DRAM_MEMORY_CONFIG
+    if output_memory_config is None:
+        output_memory_config = ttnn.DRAM_MEMORY_CONFIG
```
If any of these is ever true, this suggests a bug in `MasterConfigLoader.get_suite_parameters`, would suggest removing this validation block and the default `=None` from the parameters above (and in the previous file), and addressing any issues (assuming there are any) in the implementation of `get_suite_parameters` rather than requiring checks in `run`.

Alternatively, remove this validation block and replace the `=None` default parameters above with the respective defaults you substitute in the validation block here (omitting a default argument for `input_shape` since none is provided), if that's really the expected usage of `MasterConfigLoader`.

patrickroberts reviewed (commented):
Also, consider doing
```sh
git lfs track "model_tracer/traced_operations/ttnn_operations_master.json"
```
to avoid bloating the repository size with diffs of this file.

bbradelTT commented:
Looking at model_tracer/traced_operations/ttnn_operations_master.json around line 19304, I see

```
          {
            "UnparsedElement": {
              "error": "Expecting property name enclosed in double quotes: line 1 column 132 (char 131)",
              "element_info": "{\"arg1\": {\"MemoryConfig\":{\"memory_layout\":\"TensorMemoryLayout::WIDTH_SHARDED\",\"buffer_type\":\"BufferType::L1\",\"shard_spec\":{\"grid\":{[{\"x\":0,\"y\":0} - {\"x\":7,\"y\":0}], [{\"x\":0,\"y\":1} - {\"x\":1,\"y\":1}]},\"shape\":\"{64, 32}\",\"orientation\":\"ShardOrientation::ROW_MAJOR\"},\"nd_shard_spec\":{\"shard_shape\":[64, 32],\"grid\":{[{\"x\":0,\"y\":0} - {\"x\":7,\"y\":0}], [{\"x\":0,\"y\":1} - {\"x\":1,\"y\":1}]},\"orientation\":\"ShardOrientation::ROW_MAJOR\",\"shard_distribution_strategy\":\"ShardDistributionStrategy::ROUND_ROBIN_1D\"},\"created_with_nd_shard_spec\":0}}}"
            }
          },
          {
            "arg2": "1"
          }
        ]
      ]
    },
    "ttnn::linear": {
      "configurations": [
        [
          {
            "UnparsedElement": {
              "error": "Expecting ',' delimiter: line 1 column 359 (char 358)",
              "element_info": "{\"arg0\": {\"Tensor\":{\"storage\":{},\"tensor_spec\":{\"logical_shape\":[1, 1, 1, 8],\"tensor_layout\":{\"dtype\":\"DataType::BFLOAT8_B\",\"page_config\":{\"config\":{\"tile\":{\"tile_shape\":\"{32, 32}\",\"face_shape\":\"{16, 16}\",\"num_faces\":4}}},\"memory_config\":{\"memory_layout\":\"TensorMemoryLayout::BLOCK_SHARDED\",\"buffer_type\":\"BufferType::L1\",\"shard_spec\":{\"grid\":[{\"x\":0,\"y\":0} - {\"x\":0,\"y\":0}],\"shape\":\"{32, 32}\",\"orientation\":\"ShardOrientation::ROW_MAJOR\"},\"nd_shard_spec\":{\"shard_shape\":[32, 32],\"grid\":[{\"x\":0,\"y\":0} - {\"x\":0,\"y\":0}],\"orientation\":\"ShardOrientation::ROW_MAJOR\",\"shard_distribution_strategy\":\"ShardDistributionStrategy::GRID_2D\"},\"created_with_nd_shard_spec\":0},\"alignment\":[32, 32]}}}}}"
            }
```

Shouldn't these be parsed properly?

Aswinmcw commented:
> Looking at model_tracer/traced_operations/ttnn_operations_master.json around line 19304, I see
> 
> ```
>           {
>             "UnparsedElement": {
>               "error": "Expecting property name enclosed in double quotes: line 1 column 132 (char 131)",
>               "element_info": "{\"arg1\": {\"MemoryConfig\":{\"memory_layout\":\"TensorMemoryLayout::WIDTH_SHARDED\",\"buffer_type\":\"BufferType::L1\",\"shard_spec\":{\"grid\":{[{\"x\":0,\"y\":0} - {\"x\":7,\"y\":0}], [{\"x\":0,\"y\":1} - {\"x\":1,\"y\":1}]},\"shape\":\"{64, 32}\",\"orientation\":\"ShardOrientation::ROW_MAJOR\"},\"nd_shard_spec\":{\"shard_shape\":[64, 32],\"grid\":{[{\"x\":0,\"y\":0} - {\"x\":7,\"y\":0}], [{\"x\":0,\"y\":1} - {\"x\":1,\"y\":1}]},\"orientation\":\"ShardOrientation::ROW_MAJOR\",\"shard_distribution_strategy\":\"ShardDistributionStrategy::ROUND_ROBIN_1D\"},\"created_with_nd_shard_spec\":0}}}"
>             }
>           },
>           {
>             "arg2": "1"
>           }
>         ]
>       ]
>     },
>     "ttnn::linear": {
>       "configurations": [
>         [
>           {
>             "UnparsedElement": {
>               "error": "Expecting ',' delimiter: line 1 column 359 (char 358)",
>               "element_info": "{\"arg0\": {\"Tensor\":{\"storage\":{},\"tensor_spec\":{\"logical_shape\":[1, 1, 1, 8],\"tensor_layout\":{\"dtype\":\"DataType::BFLOAT8_B\",\"page_config\":{\"config\":{\"tile\":{\"tile_shape\":\"{32, 32}\",\"face_shape\":\"{16, 16}\",\"num_faces\":4}}},\"memory_config\":{\"memory_layout\":\"TensorMemoryLayout::BLOCK_SHARDED\",\"buffer_type\":\"BufferType::L1\",\"shard_spec\":{\"grid\":[{\"x\":0,\"y\":0} - {\"x\":0,\"y\":0}],\"shape\":\"{32, 32}\",\"orientation\":\"ShardOrientation::ROW_MAJOR\"},\"nd_shard_spec\":{\"shard_shape\":[32, 32],\"grid\":[{\"x\":0,\"y\":0} - {\"x\":0,\"y\":0}],\"orientation\":\"ShardOrientation::ROW_MAJOR\",\"shard_distribution_strategy\":\"ShardDistributionStrategy::GRID_2D\"},\"created_with_nd_shard_spec\":0},\"alignment\":[32, 32]}}}}}"
>             }
> ```
> 
> Shouldn't these be parsed properly?

It will be parsed properly when loading these vectors in sweep test

Aswinmcw commented:
> > Also, consider doing
> > ```shell
> > git lfs track "model_tracer/traced_operations/ttnn_operations_master.json"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to avoid bloating the repository size with diffs of this file.
> 
> I thought we don't use lfs any more.

The plan is to move this json to the DB rather than storing it in our metal repo

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Refactor run function parameters in model tracer and tests

### What's changed
- Updated the `run` function signatures in `TRACED_CONFIGS_GUIDE.md`, `add_all_pytorch2.py`, and `sigmoid_accurate.py` to make `input_shape` a required parameter and set default values for other parameters.

### Checklist
- Ensured consistency across all relevant files for parameter handling.

bbradelTT commented:
> > > Also, consider doing
> > > ```shell
> > > git lfs track "model_tracer/traced_operations/ttnn_operations_master.json"
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > to avoid bloating the repository size with diffs of this file.
> > 
> > 
> > I thought we don't use lfs any more.
> 
> The plan is to move this json to the DB rather than storing it in our metal repo

Will that be in a future PR?
If that is the case, maybe make it really small in this PR.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

stevendae commented on model_tracer/TRACED_CONFIGS_GUIDE.md:
```diff
@@ -0,0 +1,213 @@
+python model_tracer/generic_ops_tracer.py /path/to/model/test.py::test_function
+
+# Example
+python model_tracer/generic_ops_tracer.py models/experimental/efficientnetb0/tests/pcc/test_ttnn_efficientnetb0.py::test_efficientnetb0_model
```
Here is the output I get:

```
(python_env) stevenlee@aus-wh-20-special-stevenlee-for-reservation-48542:/localdev/stevenlee/tt-metal$ python model_tracer/generic_ops_tracer.py models/experimental/efficientnetb0/tests/pcc/test_ttnn_efficientnetb0.py::test_efficientnetb0_model
🚀 TTNN Operations Tracer
==================================================
📁 test_ttnn_efficientnetb0.py::test_efficientnetb0_model
==================================================
🚀 Running test with operations tracing...
❌ Error: [Errno 2] No such file or directory: '/home/ubuntu/tt-metal/conftest_tracer.py'
Traceback (most recent call last):
  File "/localdev/stevenlee/tt-metal/model_tracer/generic_ops_tracer.py", line 605, in main
    result = run_test_with_tracing(args.test_path, args.output_dir, args.store)
  File "/localdev/stevenlee/tt-metal/model_tracer/generic_ops_tracer.py", line 526, in run_test_with_tracing
    plugin_file = create_tracing_plugin(output_dir)
  File "/localdev/stevenlee/tt-metal/model_tracer/generic_ops_tracer.py", line 506, in create_tracing_plugin
    with open(plugin_file, "w") as f:
FileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/tt-metal/conftest_tracer.py'
```

stevendae replied:
still getting same error as above. tested from sha f9eaf99ded10c33d6ecd2d79e98842823c37b7d4

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

stevendae commented on model_tracer/analyze_sweep_results.py:
```diff
@@ -0,0 +1,334 @@
+class SweepResultsAnalyzer:
+    """Analyzes sweep test results from exported JSON files"""
+
+    def __init__(self, results_dir: str = "/home/ubuntu/tt-metal/tests/sweep_framework/results_export"):
```
this persists

Aswinmcw committed:
Refactor model_tracer scripts to use dynamic base directory resolution

- Updated `analyze_sweep_results.py` and `generic_ops_tracer.py` to determine the base directory from `PYTHONPATH` or current working directory, enhancing flexibility.
- Modified default result directory paths to be relative to the dynamically determined base directory.
- Adjusted example paths in documentation to remove hardcoded directory references.
- Updated JSON configuration files to correct parsing errors and improve structure.

This change improves the usability of the scripts across different environments and setups.

Aswinmcw committed:
Refactor base directory resolution in model_tracer and master_config_loader

- Updated `analyze_sweep_results.py` and `master_config_loader.py` to dynamically import `get_base_dir` from `generic_ops_tracer`, enhancing flexibility in locating the base directory.
- Removed hardcoded base directory logic, allowing for better adaptability across different environments.
- Ensured fallback mechanisms are in place for importing the base directory function, improving robustness.

These changes streamline the configuration process and improve the usability of the scripts in varied setups.

Aswinmcw committed:
Update sweep test command in TRACED_CONFIGS_GUIDE.md to include vector source and result destination parameters

- Modified the command for running sweep tests to specify `--vector-source` and `--result-dest` options, enhancing clarity and usability for users.
- Updated example usage in the documentation to reflect these changes, ensuring consistency and improved guidance for users executing sweep tests.

Aswinmcw committed:
Enhance MasterConfigLoader to support input deduplication for test configurations

- Updated `MasterConfigLoader` to allow deduplication of input configurations for both unary and binary operations, improving efficiency in test generation.
- Modified the `get_suite_parameters` method to include a `deduplicate_inputs` flag, which controls whether to return unique input configurations or all combinations.
- Adjusted related methods to handle deduplication logic, ensuring that only unique configurations are processed and returned.
- Updated example usages in the documentation to reflect changes in behavior for deduplicated inputs.

These enhancements streamline the testing process and reduce redundancy in generated test cases.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Enhance ttnn-run-sweeps workflow by adding support for 'ALL SWEEPS (model_traced)'. Updated conditions for sweep execution to include this new option, ensuring proper handling in both single and all sweeps generation jobs.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Revert "Enhance ttnn-run-sweeps workflow by adding support for 'ALL SWEEPS (model_traced)'. Updated conditions for sweep execution to include this new option, ensuring proper handling in both single and all sweeps generation jobs."

This reverts commit 12519e8f684ba513eda404f45f319cceb3b366b9.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Update sigmoid_accurate.py to include all cases in model tracing parameters

Aswinmcw committed:
Refactor MasterConfigLoader to return paired parameters as tuples in sigmoid_accurate.py

- Updated MasterConfigLoader to return paired parameters as tuples instead of separate config names, improving data structure for traced configurations.
- Adjusted sigmoid_accurate.py to utilize the new parameter format, ensuring compatibility with the updated loader functionality.

Aswinmcw committed:
Refactor MasterConfigLoader to improve tensor input handling

- Renamed `_is_binary_operation` to `_count_tensor_inputs` to clarify its purpose of counting tensor inputs.
- Introduced logic to handle operations with varying numbers of tensor inputs, including unary, binary, and multi-input operations.
- Updated related methods to utilize the new counting function, enhancing the clarity and maintainability of the code.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Refactor MasterConfigLoader to improve configuration handling

### What's changed
- Replaced tuple-based parameter returns with individual lists for better clarity and alignment with sweep file expectations.
- Added support for extracting operation-specific parameters, including a default `dims` parameter for the "permute" operation.
- Introduced a new method `_extract_permute_dims` to handle the extraction of `dims` from configuration.

### Impact
This refactor enhances the readability and maintainability of the configuration loading process, ensuring that operation-specific parameters are correctly handled and returned.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Add model traced sweeps for various operations and update workflow configuration

### What's changed
- Introduced new model traced sweep files for operations including add, conv2d, global_avg_pool2d, linear, multiply, pad, permute, reshape, sharded_to_interleaved, sigmoid_accurate, silu, swish, transpose, and untilize.
- Updated the GitHub Actions workflow to include a new sweep suite option for model traced operations.
- Enhanced the workflow to set a default suite name and handle suite selection based on the sweep type.

### Impact
This addition expands the testing coverage for model traced operations, ensuring better validation and performance tracking across various tensor operations in the TTNN framework.

Aswinmcw committed:
Update workflow to include 'Model Traced' sweep option for ttnn sweeps

### What's changed
- Modified conditions in the GitHub Actions workflow to incorporate 'ALL SWEEPS (Model Traced)' for both single and all sweeps generation.
- Ensured that the compute module matrix job also recognizes the new sweep option.

### Impact
This update enhances the flexibility of the sweep generation process, allowing for better testing and validation of model traced operations in the TTNN framework.

Aswinmcw committed:
Refactor add and sigmoid_accurate sweep tests to remove master config loader

### What's changed
- Removed the MasterConfigLoader and associated model traced parameters from both `add_all_pytorch2.py` and `sigmoid_accurate.py`.
- Updated the `run` function signatures to accept parameters directly instead of unpacking from traced configurations.

### Impact
This refactor simplifies the test setup by eliminating unnecessary dependencies on master configurations, enhancing clarity and maintainability of the sweep tests.

Aswinmcw committed:
Enhance MasterConfigLoader to include traced configuration names and streamline parameter handling

### What's changed
- Updated the MasterConfigLoader to generate traced configuration names for each configuration in the paired_configs.
- Refactored the parameter handling to create tuples of exact configurations, preventing Cartesian product issues.
- Simplified the addition of operation-specific parameters, including traced configuration names, to the result dictionary.

### Impact
These changes improve the clarity and maintainability of the configuration loading process, ensuring that traced configurations are correctly handled and enhancing the overall functionality of the sweep tests.

Aswinmcw committed:
Refactor ttnn sweeps generation and enhance parameter handling for model traced operations

### What's changed
- Updated the GitHub Actions workflow to separate the handling of 'Model Traced' sweeps from other sweep types, allowing for more specific execution conditions.
- Modified the `generate_tests` function in `sweeps_parameter_generator.py` to include a new `model_traced_only` parameter, enabling targeted generation of test vectors for operations in the `model_traced` directory.

### Impact
These changes improve the flexibility and clarity of the sweep generation process, ensuring that model traced operations are handled distinctly and efficiently within the testing framework.

Aswinmcw committed:
Update workflow conditions and refine MasterConfigLoader parameter handling

### What's changed
- Enhanced the GitHub Actions workflow to exclude 'ALL SWEEPS (Model Traced)' from certain execution conditions, improving control over sweep execution.
- Commented out the handling of `traced_config_name` in the MasterConfigLoader, clarifying that it is metadata only and not passed to the run function.

### Impact
These changes streamline the workflow and improve the clarity of parameter handling in the configuration loader, ensuring that only relevant parameters are processed during sweep generation.

Aswinmcw committed:
Enhance MasterConfigLoader and update model tracing for conv2d operations

### What's changed
- Added special handling for `conv2d` and `linear` operations in the MasterConfigLoader to extract and manage their parameters effectively.
- Updated the `run_collective_update.py` to include a new run type for 'model_traced', enhancing the flexibility of sweep execution.
- Integrated model traced configurations into the `conv2d_model_traced.py` to ensure valid configurations are included in the sweep parameters.

### Impact
These changes improve the handling of complex parameter structures for specific operations, streamline the configuration loading process, and enhance the testing framework's capability to manage model traced operations efficiently.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Update along with llama trace

### Changes made
- Renamed the operation from `ttnn::pad` to `ttnn::embedding` in the traced operations JSON file.
- Added new operation parameter extractors for `embedding` and other operations in the master config loader.
- Introduced new model traced sweep files for various operations including `embedding`, `concat`, `slice`, and others to enhance testing coverage.

### Impact
These changes improve the support for embedding operations and enhance the testing framework for various tensor operations, ensuring better validation and performance checks.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Enhance configuration handling in model tracing

- Updated `print_operation_configs` to support both new (dict with source) and old (list) formats for configurations, ensuring backward compatibility.
- Modified `OperationsTracingPlugin` to convert old format configurations to the new format with source tags when loading operations.
- Enhanced master JSON file handling to convert existing configurations from list to dict format with source information.
- Improved `MasterConfigLoader` to normalize configurations for backward compatibility.
- Refactored various sweep tests to accommodate changes in configuration formats and ensure proper handling of input shapes and memory configurations.

This update aims to streamline the configuration management process and improve compatibility with legacy formats.

Aswinmcw committed:
Refactor operation parameter extraction and configuration handling

- Introduced a registry-based approach for extracting operation-specific parameters in `MasterConfigLoader`, improving maintainability and extensibility.
- Updated `OperationParameterExtractors` to support new extraction methods for various operations, including `permute`, `untilize_with_unpadding`, and `nlp_concat_heads_decode`.
- Enhanced parameter transformation methods to accept optional parsing functions for data types, layouts, and memory configurations, allowing for more flexible configuration handling.
- Added a new test file for `all_gather_async` model tracing, incorporating comprehensive parameter sweeps and validation checks.

These changes aim to streamline the configuration management process and enhance the robustness of model tracing functionalities.

Aswinmcw committed:
Enhance operation parameter extraction for new operations

- Added support for extracting parameters for `paged_scaled_dot_product_attention_decode` and `all_gather_async` operations in `OperationParameterExtractors`.
- Implemented specific extraction and transformation methods to handle these operations, improving the flexibility and robustness of parameter management.
- Updated `MasterConfigLoader` to utilize the new extraction methods, ensuring proper handling of operation-specific configurations.
- Refactored related tests to validate the new extraction logic and ensure compatibility with existing configurations.

These changes aim to streamline the parameter extraction process and enhance the overall functionality of model tracing.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Update model traced sweep files with new copyright and license headers

- Replaced the existing copyright notice with the updated © 2025 Tenstorrent AI ULC in all model traced sweep files.
- Added SPDX-License-Identifier: Apache-2.0 to each file for compliance.
- Removed the auto-generated comment to streamline the file headers.

These changes ensure that all model traced sweep files reflect the current ownership and licensing information.

Aswinmcw committed:
Enhance generic_ops_tracer.py usage and update TRACED_CONFIGS_GUIDE.md

- Updated the usage instructions in `generic_ops_tracer.py` to include new command-line options: `--output-dir` and `--store`.
- Modified example commands to reflect the new options for better clarity.
- Enhanced the description of the tracer in `TRACED_CONFIGS_GUIDE.md` to reference the methodology from the graph tracing tech report.

These changes aim to improve user experience and documentation clarity for the model tracing functionality.

Aswinmcw committed:
Enhance tensor extraction and configuration handling in model tracing

- Updated `MasterConfigLoader` to improve extraction of tensor configurations from traced operations, specifically for concat operations.
- Implemented a new method in `OperationParameterExtractors` to handle extraction from `UnparsedElement`, allowing for better management of tensor vectors.
- Refactored related tests to validate the new extraction logic and ensure compatibility with existing configurations.
- Adjusted model traced sweep files to reflect updated input shape handling and added support for multiple tensor configurations.

These changes aim to streamline the parameter extraction process and enhance the overall functionality of model tracing.

Aswinmcw committed:
Update copyright and license headers in reshape_model_traced.py

- Added updated copyright notice for © 2025 Tenstorrent AI ULC.
- Included SPDX-License-Identifier: Apache-2.0 for compliance.
- Removed the auto-generated comment to streamline the file header.

These changes ensure proper ownership and licensing information in the model traced sweep files.

Aswinmcw committed:
Refactor MasterConfigLoader and operation parameter extractors for improved tensor handling

- Updated `MasterConfigLoader` to enhance deduplication logic for tensor configurations, particularly for reshape operations.
- Improved handling of input shapes and memory configurations in model tracing, ensuring compatibility with various tensor formats.
- Enhanced `OperationParameterExtractors` to better manage JSON parsing and tensor extraction, including aggressive fixes for common parsing issues.
- Adjusted model traced sweep files to reflect these changes, ensuring accurate configuration handling and improved performance.

These updates aim to streamline tensor management and enhance the robustness of model tracing functionality.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Refactor input specifications handling in MasterConfigLoader

- Updated `MasterConfigLoader` to pair input specifications with their corresponding `is_conv1d` values, preventing Cartesian product issues.
- Changed the return structure to use a comma-separated parameter name for better clarity and organization.

These changes enhance the configuration management for model tracing.

Aswinmcw committed:
Refactor global average pooling handling in model tracing

- Updated the global average pooling logic to correctly preserve the last dimension and match the expected output shape.
- Enhanced shard shape validation to ensure tile-alignment, adjusting layout as necessary.
- Added slicing of the output tensor to align with the expected shape after pooling.

These changes improve the accuracy and performance of model tracing operations.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
update

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Refactor embedding configuration handling in MasterConfigLoader

Updated the embedding configuration logic to return tuples instead of separate lists for parameters. This change prevents cartesian product explosion and keeps related configuration values together, improving clarity and maintainability. The return structure now uses a single parameter name with a tuple list for better alignment with other operations.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Enhance MasterConfigLoader and add new operation parameter extractors

- Updated MasterConfigLoader to validate sharding configurations for tile alignment and adjusted shard shapes to be tile-aligned.
- Added extraction methods for max_pool2d and upsample operations, enabling parameter extraction from JSON configurations.
- Introduced new test scripts for max_pool2d, upsample, and other operations, ensuring compatibility with traced configurations.
- Improved error handling and validation for extracted parameters in the operation parameter extractors.
- Updated existing tests to utilize the new parameter extraction methods, enhancing test coverage and reliability.

Aswinmcw committed:
Add typecast parameter extraction and update MasterConfigLoader

- Introduced a new method for extracting typecast parameters, including output_dtype, from JSON configurations.
- Enhanced MasterConfigLoader to handle typecast operation parameters, ensuring proper configuration loading.
- Updated model_traced test scripts to reflect changes in parameter handling, removing unnecessary defaults for output_dtype.
- Improved overall clarity and maintainability of configuration handling in the MasterConfigLoader.

Aswinmcw committed:
Update traced configurations and enhance MasterConfigLoader validation

- Modified TRACED_CONFIGS_GUIDE.md to reflect changes in module names for model tracing.
- Updated ttnn_operations_master.json with new tensor configurations, including adjustments to logical shapes and data types.
- Enhanced MasterConfigLoader with a new method to validate TTNN MemoryConfig objects, ensuring proper shard specifications and alignment.
- Added detailed checks for invalid configurations, reporting potential issues without filtering them out, to improve runtime reliability.
- Improved overall clarity and maintainability of configuration handling in the MasterConfigLoader.

Aswinmcw committed:
Enhance MasterConfigLoader with additional layout checks and memory configuration updates

- Added support for ROW_MAJOR layout in tilize and pad operations based on JSON configuration and padding checks.
- Implemented checks for front padding in pad operations to determine layout requirements.
- Updated memory configuration handling for upsample and untilize_with_unpadding operations to ensure proper layout alignment.
- Improved overall validation and clarity in MasterConfigLoader for operation-specific configurations.

mtairum reviewed (changes requested):
The commented out code in `models/tt_transformers/demo/simple_text_demo.py` seems to be a mistake.

gwangTT commented:
How can we remind people who work on new models to follow [the instructions here](https://github.com/tenstorrent/tt-metal/blob/Aswinmcw/model_trace/model_tracer/TRACED_CONFIGS_GUIDE.md#1-trace-a-model) to update the `model_tracer/traced_operations/ttnn_operations_master.json`?

- Maybe [the PR template](https://github.com/tenstorrent/tt-metal/blob/main/.github/pull_request_template.md) should be updated to include a checkbox for running the model_tracer?
- Another idea is to add a nightly (or weekly?) test that runs all the models in a special mode (would `TT_METAL_KERNELS_EARLY_RETURN` work for this?) to regenerate `model_tracer/traced_operations/ttnn_operations_master.json`?

stevendae commented:
> How can we remind people who work on new models to follow [the instructions here](https://github.com/tenstorrent/tt-metal/blob/Aswinmcw/model_trace/model_tracer/TRACED_CONFIGS_GUIDE.md#1-trace-a-model) to update the `model_tracer/traced_operations/ttnn_operations_master.json`?
> 
> * Maybe [the PR template](https://github.com/tenstorrent/tt-metal/blob/main/.github/pull_request_template.md) should be updated to include a checkbox for running the model_tracer?
> * Another idea is to add a nightly (or weekly?) test that runs all the models in a special mode (would `TT_METAL_KERNELS_EARLY_RETURN` work for this?) to regenerate `model_tracer/traced_operations/ttnn_operations_master.json`?

is the process for adding new model traced sweeps & parameters documented?

stevendae committed:
test(sweeps): make paged decode and tilize_with_val_padding sweeps honest

stevendae reviewed (approved):
ok to land as a first iteration of trace‑to‑sweep infra. Follow‑ups should tighten sharding fallbacks, add loader/extractor tests, and eventually refactor the extractor/loader modules for better SRP.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Update traced operations to include deepseek model traace

- Modified `ttnn_operations_master.json` to adjust tensor shapes, data types, and memory configurations for various operations.
- Enhanced `simple_text_demo.py` with additional test cases for different input scenarios, including reasoning and CI performance evaluations.
- Added new test cases in sweep framework scripts to ensure comprehensive coverage of tensor operations and configurations.

This update aims to refine the model's performance and ensure accurate testing across various configurations.

Aswinmcw committed:
Add README for Model Trace to Test Automation

- Introduced a comprehensive README.md for the Model Trace feature, detailing its purpose, benefits, and integration patterns.
- Included quick reference commands for tracing models, viewing configurations, and running sweep tests.
- Provided examples for both unary and binary operations, along with usage instructions and a complete workflow overview.
- Documented file structure and key components to facilitate user understanding and implementation.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Update README.md for Model Tracer integration

- Revised integration steps for model tracing, simplifying the process from three changes to two steps.
- Clarified the contents of the Master JSON file, removing the total config count for brevity.
- Enhanced examples for viewing configurations, directing users to specific model traced files.
- Added notes regarding the need for additional test files and custom parameter extraction for certain operations.

Aswinmcw commented:
@mtairum - can you please unblock this PR, i have reverted the change made in the models

@gwangTT - you can check the status of these sweeps in regular sweeps results dashboard - https://superset.tenstorrent.com/superset/dashboard/metal-test-runs/

Regarding notifying the model writers to run this model tracer - @stevendae can confirm how to iterate this

Running the model and sweep is straightforward:
The user only needs to run the model tracer with the model path. This will append the required configs to the master JSON file. After that, the usual sweep run will automatically load those configs.

Aswinmcw committed:
Update README.md and add TODO in paged scaled dot product attention test

- Added a blank line for better readability in the README.md.
- Updated a comment in the paged scaled dot product attention test to indicate the need for a true PyTorch attention golden output computation.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Add environment variable support for HuggingFace models and Llama directories in operations tracing

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Enhance operations tracing and configuration handling

- Added a unique test counter to the OperationsTracingPlugin to ensure trace file uniqueness.
- Implemented retry logic for loading existing master files to handle file locking issues.
- Improved error handling for trace file reading, providing clearer error messages.
- Updated various model tracing tests to include a default storage type of "StorageType::DEVICE" for better compatibility and consistency across tests.
- Adjusted tensor configuration handling to extract and utilize storage type effectively.

This update aims to enhance the robustness and clarity of the operations tracing process, ensuring smoother execution and better error reporting.

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Enhance workflow conditions and README updates

- Added conditional checks in the GitHub Actions workflow to determine if jobs should run based on the event type and sweep name.
- Updated the README in the model_tracer directory to reflect changes in command-line arguments for running the model_traced suite, improving clarity on vector source usage.

Aswinmcw committed:
Enhance vector artifact handling in GitHub Actions workflow

- Added conditional checks to ensure the artifacts directory exists and contains JSON files before attempting to merge them.
- Improved error handling and logging to provide clearer feedback when no vector files are found or when the artifacts directory is missing.
- Updated the matrix computation step to handle cases where no vector files are available, ensuring robustness in the workflow execution.

Aswinmcw committed:
Update GitHub Actions workflow for sweep conditions and error handling

- Enhanced the conditions for job execution in the workflow to include "ALL SWEEPS (Model Traced)" in the manual dispatch entry.
- Improved error messaging when no vector files are found, providing clearer guidance for troubleshooting in the event of a failure during the vector generation step.

Aswinmcw committed:
Refine GitHub Actions workflow for sweep execution and artifact handling

- Updated the conditions for job execution to ensure proper handling of the "ALL SWEEPS (Model Traced)" case.
- Enhanced error handling to provide clearer messages when the artifacts directory is missing or contains no vector files, improving troubleshooting guidance.
- Improved the logic for moving JSON files from the artifacts directory to ensure robustness in the workflow.

Aswinmcw committed:
Refactor GitHub Actions workflow for sweep execution

- Removed unnecessary conditions for job execution, simplifying the logic for manual dispatch and scheduled runs.
- Enhanced clarity in comments regarding the sweep execution entries.
- Improved artifact handling by streamlining the process of moving JSON files and removing redundant error checks.
- Updated the suite name handling to ensure it reflects the input correctly without defaulting to "nightly".

Aswinmcw committed:
Merge branch 'main' into Aswinmcw/model_trace

Aswinmcw committed:
Update ttnn-run-sweeps.yaml to add a new cron schedule for daily model traced runs at 4:30 AM UTC. Adjust conditions for running sweeps generation based on event type and schedule. Modify device fixture in device_fixtures.py to simplify device initialization.

Aswinmcw committed:
Refine conditions for running ttnn sweeps generation in workflow. Update logic to ensure sweeps are triggered correctly based on event type and input parameters.

Aswinmcw committed:
Add model traced run type to ttnn sweeps workflow. Update conditions to trigger model traced sweeps based on specific schedule and input parameters.
